OpenSC/OpenSCToken.cpp

```diff
@@ -327,33 +445,141 @@ char tokenUid[TOKEND_MAX_UID])
+                        
+                        n = 0; // to mark that we haven't retrieved the commonName yet
+                        
+                        // Convert this cert from DER to internal representation (structure)
+                        x = d2i_X509(NULL, (const u8 **)&cert_der, cert_der_len);
+                        free(cert_der2); cert_der = cert_der2 = NULL; // free on the spot to avoid potential memory leak
+                        
+                        if (x == NULL) {
+                                sc_debug(mScCtx, SC_LOG_DEBUG_NORMAL, "      failed to parse ASN.1 X509 cert...\n");
+                                goto end;
+                        }
+                        
+                        // Get X509_NAME construct pointer (internal structure - must not be freed!)
+                        x509_name = X509_get_subject_name(x);
+                        
+                        // Determine how long commonName is
+                        n = X509_NAME_get_text_by_NID(x509_name, NID_commonName, NULL, 0);
+                        if (n <= 0) {
+                                sc_debug(mScCtx, SC_LOG_DEBUG_NORMAL, "      failed to get good commonName length (%d)\n", n);
+                                goto end;
+                        } else {
+                                n = 0; // we don't need it really, as our allocations are based on PATH_MAX
+                        }
+                        
+                        
+                        // Extract X509 commonName in ASCII form
+                        subject = (char *) malloc(PATH_MAX);
+                        if (subject == NULL) goto end;
+                        memset(subject, 0, PATH_MAX);
+                        
+                        n = X509_NAME_get_text_by_NID (x509_name, NID_commonName, subject, PATH_MAX - 1);
+                        
+                        // And place that name where Tokend should pick it for display
+                        strlcpy(printName, (const char *) subject, PATH_MAX);
+                        
+                        sc_debug(mScCtx, SC_LOG_DEBUG_NORMAL, "      printName (%d bytes): \"%s\"\n", n, printName);
+                        
+                end:
+                        if (subject != NULL) {
+                                free(subject); subject = NULL;
+                        }
+                        if (x != NULL) {
+                                OPENSSL_free(x); x = NULL;
+                        }
+                        
+                        // We got our subject->commonName, so no need to repeat the same
+                        // for all the certificates on this token (one is enough)
+                        if (n > 0) // we got our printName
+                                break;
+                }
+	
+	}
+		
+	// from PUBKEY I can learn whether it is ECC or RSA.
+	r = sc_pkcs15_get_objects(mScP15Card, SC_PKCS15_TYPE_PUBKEY, objs, 32);
```

Comment:
Use SC_PKCS15_TYPE_PRKEY, some tokens do not have PKCS11 public key object
and then compare
if (objs[0]->type == SC_PKCS15_TYPE_PRKEY_EC)